README.md

```diff
@@ -0,0 +1,536 @@
+
+## MQTT Protocol
+
+[MQTT](https://mqtt.org/) is a messaging protocol designed specifically for (IoT) device
+communications. It is used in many areas, such as banking, oil & gas industry, manufacturing, etc.
+
+MQTT protocol has many advantages, but here I want to mention some of them:
+
+* It is a lightweight binary protocol generally running over TCP/IP.
+* It is designed for unreliable networking, making it a good choice for outdoor installations.
+* It follows pub/sub model simplifying client logic.
+
+We will demonstrate some of MQTT's advantages in our setup.
+
+## MQTT Brokers
+
+An essential thing about MQTT is that it simplifies the client's logic, which is critical for embedded devices.
+This is achieved with the pubsub model: in MQTT, there is no concept of a "server".
+Instead, all participating entities are clients connecting to a so-called _broker_.
+Clients _subscribe_ to _topics_, _publish_ messages to them, and the broker does the routing (and many other things).
+
+A good production-ready broker, like [EMQX](https://github.com/emqx/emqx) generally provides
+not only MQTT routing capabilities but many other intellectual features, like
+* other kinds of connection methods, like WebSockets;
+* different models of authentication and authorization;
+* streaming data to databases;
+* custom routing rules based on message introspection;
+* and so on.
+
+## Sensor Setup
+
+For the sake of simplicity, our device will be represented by an ordinary Mix
+application: it can be easily converted to a Nerves application.
+
+First, we create a Mix project:
+
+```bash
+mix new --sup weather_sensor
+cd weather_sensor
+```
+
+To interact with an MQTT broker, we need an MQTT client. We take [emqtt](https://github.com/emqx/emqtt).
+Add it to `mix.exs` as a dependency:
+
+```elixir
+defp deps do
+  [
+    {:emqtt, github: "emqx/emqtt", tag: "1.4.4"}
+  ]
+end
+```
+
+```bash
+# we do not want to use and build QUIC capabilities
+export BUILD_WITHOUT_QUIC=1
+mix deps.get
+```
+
+We will put all our "sensor" code to the main module `WeatherSensor`, so we need to
+add it to the application supervisor `lib/weather_sensor/application.ex`:
+
+```elixir
+defmodule WeatherSensor.Application do
+  @moduledoc false
+
+  use Application
+
+  @impl true
+  def start(_type, _args) do
+    children = [
+      WeatherSensor
+    ]
+
+    opts = [strategy: :one_for_one, name: WeatherSensor.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+
+Now, let's implement the main module in `lib/weather_sensor.ex`:
+
+```elixir
+defmodule WeatherSensor do
+  @moduledoc false
+
+  use GenServer
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  def init([]) do
+    interval = Application.get_env(:weather_sensor, :interval)
+    emqtt_opts = Application.get_env(:weather_sensor, :emqtt)
+    # This is the topic to which we will publish our reports
+    report_topic = "/reports/#{emqtt_opts[:clientid]}/temperature"
```

Comment:
Nit: maybe our topics shouldn't start with a [leading slash](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/) (there's this one and a few below).

Reply:
Yes, leading slashes are discouraged